src/detect.ts

```diff
@@ -9,8 +9,23 @@ const LOCKS: Record<string, PackageManager> = {
   'package-lock.json': 'npm',
 }
 
+/**
+ * Detect the package manager and version based on the process.env var
+ * "npm_config_user_agent".
+ * Falls back to detection by lockfile without version if process.env
+ * var is somehow missing.
+ */
 export async function detectPackageManager(cwd = process.cwd()) {
-  const result = await findUp(Object.keys(LOCKS), { cwd })
-  const agent = (result ? LOCKS[path.basename(result)] : null)
-  return agent
+  let agent: string | null = null
+  let version: string | null = null
+
+  if (process.env?.npm_config_user_agent)
+    [agent, version] = process.env.npm_config_user_agent.match(/[^\/\s]+/g) ?? []
```

Comment:
I am not so sure about this check. If you have `npm run xxx` in your scripts but using yarn or pnpm, it might be inferred to `npm` instead I guess?

Reply:
You are right, I did not think about this case. I can change the pr back to detecting the package manager only by its lockfile.
Then we can decide to detect the yarn version by one of this methods:

1. check for "packageManager" entry in package.json
2. check for .yarn folder
3. stay with env var "npm_config_user_agent"
4. run yarn -v command

1. or 2. may not work in every case (e.g. in yarn workspaces when you run the command inside a nested folder)
3. Only works if you are running the current script with yarn. In case you run npm but you have a yarn lockfile you won't have a version of yarn.
4. Should always work

So I would probably stick with 4.

Reply:
I would prefer 1 + 2.

4 might be too much imo.